pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -2,21 +2,34 @@ package reconciler
 
 import (
 	"context"
-
-	"github.com/go-logr/logr"
+  
+  "github.com/go-logr/logr"
 	"github.com/pkg/errors"
 	rufiov1alpha1 "github.com/tinkerbell/rufio/api/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
 	c "github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/constants"
 	"github.com/aws/eks-anywhere/pkg/controller"
 	"github.com/aws/eks-anywhere/pkg/controller/clientutil"
 	"github.com/aws/eks-anywhere/pkg/controller/clusters"
 	"github.com/aws/eks-anywhere/pkg/providers/tinkerbell"
 	"github.com/aws/eks-anywhere/pkg/providers/tinkerbell/hardware"
 )
 
+const (
+	// ReconcileNewCluster means need to create nodes for a new cluster.
+	ReconcileNewCluster clusterChange = "create-new-cluster"
+	// NeedNewNode means node has k8s version change.
+	NeedNewNode clusterChange = "k8s-version-change"
+	// NeedMoreOrLessNode means node group scales up or scales down.
+	NeedMoreOrLessNode clusterChange = "scale-update"
+)
+
```

Comment:
We're trying to avoid "operations/events" and address difference between desired and current state.
https://github.com/aws/eks-anywhere/pull/4971#discussion_r1113709519

Reply:
We're determining what the user is trying to do so we can enforce appropriate constraints (namely preventing a K8s version upgrade concurrently with a scale). All the while we impose the constraint the concept of an `Operation` will be necessary. Naming it something else, which is arguably more ambiguous, doesn't change that fact.

If we want to model current and desired state only it will be necessary to remove the concurrent operation constraint and I'm not sure we want to do that.